pkg/backend/stack_outputs.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/pulumi/pulumi/pkg/v3/secrets"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+)
+
+type stackOutputsProvider struct {
+	delegate secrets.Provider
+
+	lastDecryptError error
+}
+
+var _ secrets.Provider = (*stackOutputsProvider)(nil)
+
+func newStackOutputsProvider(delegate secrets.Provider) *stackOutputsProvider {
+	return &stackOutputsProvider{
+		delegate: delegate,
+	}
+}
+
+func (p *stackOutputsProvider) Error() error {
+	return p.lastDecryptError
+}
+
+func (p *stackOutputsProvider) OfType(ty string, state json.RawMessage) (secrets.Manager, error) {
+	delegateManager, err := p.delegate.OfType(ty, state)
+	if err != nil {
+		return nil, err
+	}
+
+	delegateDecrypter := delegateManager.Decrypter()
```

Comment:
Why are we doing this here? I don't think we need to grab the decrypter this early.

Reply:
Was more thinking neater to "know" it's non-`nil` later on, and all our `Decrypter()` calls happen to be just property accesses right now. But I can push it down :+1: